### PR TITLE
Add DJ features to cue points.

### PR DESCRIPTION
**Pull request by aleksanb**
One can now halve and double the loop length with the press of a button!
Remix your prods live infront of a live audience, watch them go wild!

### CODE REVIEW

**Comment by iver56** [comment]
How about being able to drag the loop points (and have them snap to BEATs as you drag) instead? That would be more useful, IMO. I need to slightly adjust my selected loop more often than I need to double or halve its length. I wouldn't use nin as a DJ tool.

**Comment by aleksanb** [comment]
It's quite fun to play with halving and doubling the loop length.
You should add dragging of loop points in addition, as both features are pretty cool.

**Comment by sigvef** [comment]
Consider fixing linting errors:

image

**Comment by aleksanb** [comment]
No lint settings defined in the project, we'll take it as we get that merged in?

File: nin/frontend/app/scripts/controllers/bottom.js [hunk 1]
```diff
@@ -72,6 +72,17 @@
...
 
+      commands.on('multiplyLoopLengthBy', function(amount) {
+        if ($scope.loopEnd === undefined || $scope.loopStart === undefined) {
+          return;
+        }
+
+        var clampedAmount = Math.max(0, amount),
+            loopLength = $scope.loopEnd - $scope.loopStart;
+
+        $scope.loopEnd = $scope.loopStart + (loopLength * clampedAmount);
```
Comment (sigvef) [inline nin/frontend/app/scripts/controllers/bottom.js]: Perhaps clamp the new loopLength to at least 1, so we wont get a loop that starts and ends on the same frame?

**Comment by aleksanb** [comment]
CLAMPD

File: nin/frontend/app/scripts/controllers/bottom.js [hunk 2]
```diff
@@ -72,6 +72,17 @@
...
 
+      commands.on('multiplyLoopLengthBy', function(amount) {
+        if ($scope.loopEnd === undefined || $scope.loopStart === undefined) {
+          return;
+        }
+
+        var clampedAmount = Math.max(0, amount),
+            loopLength = $scope.loopEnd - $scope.loopStart;
+
+        $scope.loopEnd = Math.max(1, $scope.loopStart + (loopLength * clampedAmount));
```
Comment (sigvef) [inline nin/frontend/app/scripts/controllers/bottom.js]: This clamp may not do what you think it does! As it is written, loops cannot end on frame 0. You may have intended to stop loops from starting and ending of the same frame, i.e. a loop length of 0. In that case, you should use something like loopStart + max(1, loopLength * clampedAmount) instead.

**Comment by aleksanb** [comment]
PING

**Comment by cristeahub** [comment]
LGTM, but I think the musicians should double check this before merging.

**Comment by lionleaf** [comment]
Set an upper limit to the loop length   ( = length of music )

Otherwise works well with newest master.

**Comment by lionleaf** [comment]
Also add hotkeys to README!

**Pull request merged**